### PR TITLE
GridView auto-scrolling when navigating with keyboard

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

When a GridView is navigated using the keyboard, scroll it to display the newly selected entry.

https://www.pivotaltracker.com/story/show/182593635

#### Visuals

See below for a video demonstrating automatic scrolling of the GridView when arrow keys are pressed on the keyboard. The first video below shows the default scrolling behavior in a GridView without headers.

Note:
 - When the Grid View is scrolled, the mouse hover highlight moves away from the mouse position. This is not a new regression, the behavior is the same in the `develop` branch (e.g. when scrolling using the mouse).

https://user-images.githubusercontent.com/273837/190183984-91f7808c-3606-43f8-bcda-ac4d5f84e00f.mov

The video below shows the behavior in a GridView with headers when the GridView is first scrolled to its top-left corner. The following guidance from the Design Doc is enabled and showcased:

> Users can change the selected component by pressing the arrow keys. The Focus does not move up if it does not have to (in most cases, the focus is located in the second row from the bottom). Instead, the component list scrolls down if there are enough entries.

https://user-images.githubusercontent.com/273837/189151546-e50aaf22-6f4d-41cb-809f-60038305745f.mov

The next video shows the behavior in the same GridView as in the previous example when the GridView is first scrolled away from any of its boundary entries. Notably, scrolling happens only when the selection is moved using the keyboard arrow keys, not when changing the selection using the mouse. This behavior is based on a discussion with @wdanilo on Discord.

https://user-images.githubusercontent.com/273837/189151974-d992be93-f61f-4e9f-9f4c-dfe260bbec5b.mov

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

[ci no changelog needed]

### CODE REVIEW

File: lib/rust/ensogl/component/grid-view/src/scrollable.rs [hunk 1]
```diff
@@ -98,19 +115,41 @@ impl<InnerGridView> GridViewTemplate<InnerGridView> {
...
+            );
+            entry_selected_by_input_move <= base_grid.entry_selected.sample(&input_move_selection);
+            let scroll_margins = &frp.set_preferred_margins_around_entry;
+            _eval <- entry_selected_by_input_move.map2(scroll_margins,
+                f!([base_grid, area] ((row, col), margins) {
+                    let scroll_to = base_grid.viewport_containing_entry(*row, *col, *margins);
+                    area.scroll_to_y(-scroll_to.y);
+                    area.scroll_to_x(scroll_to.x);
+                })
+            );
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/scrollable.rs]: scroll_to_y and scroll_to_x are FRP inputs, so w should have `scroll_to` value and then assign it, properly mapped, using `<+`.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 2]
```diff
@@ -587,6 +600,48 @@ where
...
+        let max_viewport_left = left - margins.left;
+        let new_viewport_left = if viewport.right < min_viewport_right {
+            let viewport_width = viewport.size().x;
+            min_viewport_right - viewport_width
+        } else if viewport.left > max_viewport_left {
+            max_viewport_left
+        } else {
+            viewport.left
+        };
+        Vector2(new_viewport_left, new_viewport_top)
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: This function deserves detailed testing. To make it simpler, you can extract it to a separate function not taking self (only all variables needed for computation) - similarly to the `entry::visible::position` method. You can put it there: in `entry::visible` module.

**Review by farmaazon — CHANGES_REQUESTED** [review]

**Comment by akavel** [comment]
(Changed this PR back to a Draft until I add a new video showcasing the scrolling behavior without margins. I will work on this after I address code-review comments in PR #3657.)

**Comment by akavel** [comment]
(Reopened; I assume this being closed by @MichaelMauderer was by accident, as I don't see any comment with an explanation for the reason.)

File: lib/rust/ensogl/component/grid-view/src/entry/visible.rs [hunk 3]
```diff
@@ -162,3 +164,109 @@ pub fn set_position<E: display::Object>(
...
+/// In the picture below, the dashed border represents the viewport, while the solid border
+/// represents the entry contained in the viewport.
+/// ```text
+/// ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
+/// ┆                     ▲             ┆
+/// ┆         margins.top │             ┆
+/// ┆                     ▼             ┆
+/// ┆ margins ┌───────────────┐ margins ┆
+/// ┆  .left  │  entry        │  .right ┆
+/// ┆ ◀─────▶ └───────────────┘ ◀─────▶ ┆
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/entry/visible.rs:202]:
```suggestion
/// ┆  .left  │  entry        │  .right ┆
/// ┆ ◀───▶ └───────────────┘ ◀───▶ ┆
```

**Review by farmaazon — APPROVED** [review]

File: lib/rust/ensogl/component/grid-view/src/entry/visible.rs (hunk 3 above)
Comment (akavel) [reply lib/rust/ensogl/component/grid-view/src/entry/visible.rs:202]: As discussed, seems to be a fonts rendering glitch on Adam's side, to be ignored.

**Pull request merged**